File: prj3_api/src/models/document.model.ts
import mongoose from "mongoose";

const model = mongoose.model(
    "Document",
    new mongoose.Schema({
        name: { type: String, required: true },
        originalName: { type: String, required: true },
        size: Number,
        vote: {
            type: typeof new mongoose.Schema({
                detail: [{
                    author: { type: mongoose.Schema.Types.ObjectId, ref: 'Student' },
                    vote: Number
                }],
                count: Number
            }),
            default: {detail:[], count:0}
        },
        author: { type: mongoose.Schema.Types.ObjectId, ref: 'Student' },
        tags: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Tag' }],
        createdAt: Number,
        updatedAt: Number
    }, {
        timestamps: {
            currentTime: () => Date.now()
        }
    })
);

export const Document = model;

const model2 = mongoose.model(
    "Tag",
    new mongoose.Schema({
        name: { type: String, required: true },
        description: { type: String, required: true }
    })
)

export const Tag = model2;

File: prj3_app/lib/models/document.model.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:prj3_app/helpers/converter.dart';

part "document.model.freezed.dart";
part "document.model.g.dart";

@freezed
class Document with _$Document {
  @JsonSerializable(explicitToJson: true)
  const factory Document({
    @JsonKey(name: "_id") required String id,
    required String name,
    required String originalName,
    required int size,
    required Vote vote,
    required Map<String, dynamic> author,
    required List<Tag> tags,
    @DateTimeConverter() DateTime? createdAt,
    @DateTimeConverter() DateTime? updatedAt,
  }) = _Document;

  factory Document.fromJson(Map<String, dynamic> json) => _$DocumentFromJson(json);
}

@freezed
class Vote with _$Vote {
  @JsonSerializable(explicitToJson: true)
  const factory Vote({required List<VoteDetail> detail, required int count}) = _Vote;

  factory Vote.fromJson(Map<String, dynamic> json) => _$VoteFromJson(json);
}

@freezed
class VoteDetail with _$VoteDetail {
  @JsonSerializable(explicitToJson: true)
  const factory VoteDetail({required String author, required int vote}) = _VoteDetail;

  factory VoteDetail.fromJson(Map<String, dynamic> json) => _$VoteDetailFromJson(json);
}

@freezed
class Tag with _$Tag {
  @JsonSerializable(explicitToJson: true)
  const factory Tag(
      {@JsonKey(name: "_id") required String id, required String name, String? description}) = _Tag;

  factory Tag.fromJson(Map<String, dynamic> json) => _$TagFromJson(json);
}
